massa-graph/src/block_graph.rs

```diff
@@ -2678,6 +2675,59 @@ impl BlockGraph {
             })
     }
 
+    /// get the latest blockclique (or final) block ID at a given slot, if any
+    pub fn get_latest_blockclique_block_at_slot(&self, slot: &Slot) -> Option<BlockId> {
+        let (mut latest_final_block_id, mut latest_final_block_period) = self
+            .latest_final_blocks_periods
+            .get(slot.thread as usize)
+            .unwrap_or_else(|| panic!("unexpected not found latest final block period"));
+
+        let (blocks_in_thread, block_periods) = {
+            let storage_read = self.storage.read_blocks();
+            let blocks_in_thread = match storage_read.get_blocks_by_thread(slot.thread) {
+                Some(v) => v.clone(),
+                None => return None,
+            };
+            let block_periods: HashMap<BlockId, u64> = blocks_in_thread
+                .iter()
+                .map(|id| {
+                    (
+                        *id,
+                        storage_read
+                            .get(id)
+                            .expect("unexpected missing block in storage")
+                            .content
+                            .header
+                            .content
+                            .slot
+                            .period,
+                    )
+                })
+                .collect();
+            (blocks_in_thread, block_periods)
+        };
+
+        // search for the block in the blockclique
+        blocks_in_thread
+            .intersection(
+                &self
+                    .max_cliques
+                    .iter()
+                    .find(|c| c.is_blockclique)
+                    .expect("expected one clique to be the blockclique")
+                    .block_ids,
+            )
+            .for_each(|id| {
+                let period = *block_periods.get(id).unwrap();
+                if period > latest_final_block_period && period < slot.period {
+                    latest_final_block_period = period;
+                    latest_final_block_id = *id;
+                }
+            });
+
+        Some(latest_final_block_id)
+    }
```

Comment:
A more efficient way (that doesn't require adding an index to storage and iterating over all the blocks of that thread, nor locking storage at all) would be:
* iterate over every block B in the blockclique:
  * get `self.block_statuses.get(B)`
    * if the block is not Active, panic
    * if the block is final, panic
    * if the block is not in the target slot's thread, continue
    * if the block's period is equal or higher than the target slot's period, continue
    * if the block's period is strictly higher than latest_final_block_period:
      * set latest_final_block_period = B's period
      * set latest_final_block_id = B

Reply:
great!